adapters/rest-vertx/src/main/java/org/eclipse/hono/adapter/rest/VertxBasedRestProtocolAdapter.java

```diff
@@ -426,13 +426,17 @@ private void doUploadTelemetryData(final RoutingContext ctx) {
             hono.getOrCreateTelemetrySender(tenant, createAttempt -> {
                 if (createAttempt.succeeded()) {
                     TelemetrySender sender = createAttempt.result();
-                    boolean accepted = sender.send(deviceId, payload.getBytes(), contentType);
-                    if (accepted) {
-                        ctx.response().setStatusCode(HTTP_ACCEPTED).end();
-                    } else {
-                        // we currently have no credit for uploading data to Hono's Telemetry endpoint
-                        serviceUnavailable(ctx.response(), 2);
-                    }
+                    // sending message only when the "flow" is handled and credits are available
+                    // otherwise send will never happen due to no credits
+                    sender.sendQueueDrainHandler(v -> {
```

Comment:
I think we should first check, if we do have credit (already) using `sender.sendQueueFull()`. If we do have credit, send immediately and only if we don't, register the handler for deferred sending. However, in the latter case, how long do we want to wait for credit to arrive?